lib/mli_parser.ml

```diff
@@ -0,0 +1,179 @@
+open! Compat
+
+module Code_block = struct
+  type t = { location : Odoc_model.Location_.span; contents : string }
+end
+
+(* drop_first_and_last [1; 2; 3; 4] = [2; 3]. *)
+let drop_first_and_last list = List.tl list |> List.rev |> List.tl |> List.rev
+
+let slice file_contents ~(start : Odoc_model.Location_.point)
+    ~(end_ : Odoc_model.Location_.point) =
+  let lines_to_include =
+    String.split_on_char '\n' file_contents
+    |> List.mapi (fun i line -> (succ i, line))
+    |> List.filter (fun (lnum, _) -> start.line <= lnum && lnum <= end_.line)
+    |> List.map snd
+  in
+  match lines_to_include with
+  | [] -> ""
+  | [ line ] -> String.sub line start.column (end_.column - start.column)
+  | list ->
+      (* Imagine we were slicing the file from (Line 2, Column 3) to (Line 6, Column 7):
+
+       0123456789
+       1 ----------
+       2 ---[---
+       3 ---------
+       4 --
+       5 ----------
+       6 -------]--
+       7 ----------
+       8 ----------
+
+       The code below handles this multiline case, concatenating the included substrings
+       from lines 2-6 ([lines_to_include]). *)
+      let first_line = List.hd list in
+      let last_line = List.rev list |> List.hd in
```

Comment:
This can be computed in a safe way by doing this before the match on `lines_to_include`:

```ocaml
let drop_last lst =
  match List.rev lst with
  | [] -> None
  | last :: rev_tl -> Some (List.rev rev_tl, last)

let drop_first_and_last = function
  | [] -> None
  | first :: tl -> Some (first, drop_last tl)
```

```ocaml
  match drop_first_and_last lines_to_include with
  | None-> ""
  | Some (line, None) -> String.sub line start.column (end_.column - start.column)
  | Some (first_line, Some (stripped, last_line)) ->
```

Reply:
Quite right -- I've taken your suggestion, which is much better.